### PR TITLE
Set up Groovy compilation performance tests.

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

Previously we had only performance test scenarios for Java compilation. This PR parameterize performance template generation so that we can generate projects with Groovy source code. Also, it sets up `GroovyABIChangePerformanceTest`/`GroovyNonABIChangePerformanceTests`.

### CODE REVIEW

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy [hunk 1]
```diff
@@ -59,6 +61,15 @@ abstract class FileContentGenerator {
         }
         ${dependenciesBlock('api', 'implementation', 'testImplementation', subProjectNumber, dependencyTree)}             
 
+        allprojects {
+            dependencies{
+        ${
+            language == Language.GROOVY ? directDependencyDeclaration('implementation', 'localGroovy()') : ""
```
Comment (wolfs) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy]: Maybe we should use a different `Groovy` version that `localGroovy`. I think nobody who does Groovy development uses `localGroovy`. WDYT?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorConfiguration.groovy [hunk 2]
```diff
@@ -41,5 +50,89 @@ class TestProjectGeneratorConfiguration {
...
+    ])
+class TestProjectGeneratorConfigurationBuilder {
+    TestProjectGeneratorConfigurationBuilder(String projectName, Language language = Language.JAVA) {
+        this(projectName, projectName, language)
+    }
+
+    TestProjectGeneratorConfigurationBuilder(String projectName, String templateName, Language language = Language.JAVA) {
+        this.projectName = projectName
+        this.templateName = templateName
+        this.language = language
```
Comment (wolfs) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorConfiguration.groovy]: Should we provide `language` via the builder as well? What things do we need to pass to the constructor and what via the builder?

**Comment by wolfs** [comment]
We probably need tests with and without compile avoidance: So we don't make Groovy compilation much slower than it is now. Though we also want to show that non-ABI change is much faster with compile avoidance.

We probably should also test the fully up-to-date build.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorConfiguration.groovy (hunk 2 above)
Comment (blindpirate) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorConfiguration.groovy]: It's intentional to make `language` a constructor parameter: 

```
    private String productionFile(int project = 0, int pkg = 0, int file = 0) {
        if (project >= 0) {
            "project${project}/src/main/${this.language.name}/org/gradle/test/performance/${this.templateName.toLowerCase()}/project${project}/p${pkg}/Production${file}.${this.language.name}"
        } else {
            "src/main/${this.language.name}/org/gradle/test/performance/${this.templateName.toLowerCase()}/p${pkg}/Production${file}.${this.language.name}"
        }
    }
```

This method depends on `language` so we must initialize `language` in constructor and not change it any more.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy (hunk 1 above)
Comment (blindpirate) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy]: Well, I did. When I developed some little hobby project, I always use `localGroovy()`. I guess the real problem is, if we're comparing two Gradle versions, the potential Groovy version difference between these Gradle versions might interfere the performance test.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorConfiguration.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorConfiguration.groovy]: It surprises me that the (mutable) builder doesn't finalize all its state in the `create` method and requires some immutable state already.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/FileContentGenerator.groovy]: Yeah, that is a good reason to fix the Groovy version for this performance test. Then we need to remember to update the Groovy version from time to time.

**Comment by wolfs** [comment]
Looks like it is still not enough memory: https://builds.gradle.org/repository/download/Gradle_Check_IndividualPerformanceScenarioWorkersLinux/23342121:id/failure-logs/performance/build/tmp/performance-test-files/JavaABIChangePerformanceTest/assemble_for_abi_ch...vyProject/jobcm/current___________/log.txt

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/JavaTestProject.groovy [hunk 3]
```diff
@@ -41,8 +41,8 @@ enum JavaTestProject {
     LARGE_MONOLITHIC_GROOVY_PROJECT(new TestProjectGeneratorConfigurationBuilder("largeMonolithicGroovyProject", Language.GROOVY)
         .withSourceFiles(50000)
         .withSubProjects(0)
-        .withDaemonMemory('2g')
-        .withCompilerMemory('4g')
+        .withDaemonMemory('3g')
+        .withCompilerMemory('6g')
```
Comment (wolfs) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/JavaTestProject.groovy]: Is the problem the daemon memory or the compiler memory? My guess would be that the project doesn't need so much daemon memory....

**Comment by blindpirate** [comment]
Thanks @wolfs !

**Pull request merged**